core/signer/signer.go

```diff
@@ -63,16 +63,78 @@ func New(realm string,
 	}
 }
 
-// StartSigning advance epochs once per interval.
-func (s *Signer) StartSigning(ctx context.Context, interval time.Duration) {
-	for range time.NewTicker(interval).C {
-		if err := s.CreateEpoch(ctx); err != nil {
+// StartSigning advance epochs once per minInterval, if there were mutations,
+// and at least once per maxElapsed minIntervals.
+func (s *Signer) StartSigning(ctx context.Context, minInterval, maxInterval time.Duration) {
+	// Fetch last time from previous map head (as stored in the map server) if it
+	// exists:
+	var last time.Time
+	rootResp, err := s.tmap.GetSignedMapRoot(ctx, &trillian.GetSignedMapRootRequest{
+		MapId: s.mapID,
+	})
+	if err != nil {
+		glog.Infof("GetSignedMapRoot failed: %v", err)
+		last = time.Now()
+	} else {
+		mapRoot := rootResp.GetMapRoot()
+		if mapRoot != nil {
+			last = time.Unix(0, mapRoot.TimestampNanos)
+		} else {
+			last = time.Now()
+		}
+	}
+	// Start issuing epochs:
+	tc := time.Tick(minInterval)
+	for f := range genEpochTicks(sysClock{}, last, tc, minInterval, maxInterval) {
+		if err := s.CreateEpoch(ctx, f); err != nil {
 			glog.Fatalf("CreateEpoch failed: %v", err)
 		}
 	}
 }
 
-// newMutations returns a list of mutations to process and highest sequence number returned.
+type clock interface {
+	Now() time.Time
+	Since(time.Time) time.Duration
+}
+
+type sysClock struct{}
```

Comment:
I would prefer to keep "real" packages clean from mocking/testing/fake objects if possible. But I also just found out that trillian already has such an [interface](https://github.com/google/trillian/blob/8b1388864d35cd200bf9dbf0278eb64472e5417b/util/timesource.go#L24) and it contains both a [SystemTimeSource](https://github.com/google/trillian/blob/8b1388864d35cd200bf9dbf0278eb64472e5417b/util/timesource.go#L30) as well as [FakeTimeSource](https://github.com/google/trillian/blob/8b1388864d35cd200bf9dbf0278eb64472e5417b/util/timesource.go#L39). Shouldn't we reuse this one instead?